src/charm.py

```diff
@@ -48,66 +69,68 @@ def _get_pebble_config(self, event):
         }
         return pebble_config
 
-    def _on_config_changed(self, event):
-        """Handle config-changed event."""
+    def _on_config_changed(self, event: charm.ConfigChangedEvent) -> None:
+        """Handle config-changed event.
 
-        if not self._is_valid_config():
-            # Charm will be in blocked status.
+        Args:
+            event: The information about the event.
+        """
+        # Check for container connectivity
+        if not self.unit.get_container(self.service_name).can_connect():
+            event.defer()
             return
 
-        pebble_config = self._get_pebble_config(event)
-        container = self.unit.get_container(self.service_name)
+        # Check whether configuration is valid
+        config_valid, message = self._is_valid_config()
+        if not config_valid:
+            logger.info(message)
+            self.unit.status = model.BlockedStatus(message)
+            return
 
+        # Add any newly required or update any changed services
+        pebble_config = self._get_pebble_config()
+        container = self.unit.get_container(self.service_name)
         services = container.get_plan().to_dict().get("services", {})
         if services != pebble_config["services"]:
-            logger.debug("About to add_layer with pebble_config:\n{}".format(yaml.dump(pebble_config)))
+            logger.debug(f"About to add_layer with pebble_config:\n{yaml.dump(pebble_config)}")
```

Comment:
We shouldn't use format strings in logging messages (I realise it was doing this before as well). See https://juju.is/docs/sdk/styleguide#heading--templating.

Reply:
Yes, I actually had fixed this in the next PR

Reply:
Fixed